### PR TITLE
Implement Pay(All)Sui in RPC & CLI

### PR DESCRIPTION
**Pull request by gegaowp**
Testing via CLI
```
target/debug/sui client pay_all_sui --recipient 0xb92efc6cc3306e52ea8c9e4227fac0470e199adc --input-coins 0xb5eeef175071df6c46efef00b1a56b0fd0d9f8fa 0xdcc4e2369070e03092b9ebb57064c17891a6c84e --gas-budget 10000
----- Certificate ----
Transaction Hash: 4EuOjW2DnXuzA56J0biHb+Vv+tCtzCaavjdOM9RceU8=
Transaction Signature: AA==@hZy9rSAktkS5CeLoN/3gReBmmY13jEgbSBFoU26t/Vklhkvt+N5mO5t/ACtqBheTd9AJVBlUBWz7dhZglks/Bg==@xfrRTtxCS07XSV3guhSMkF0a5hw8mIKPPX+LXe7WUZQ=
Signed Authorities Bitmap: RoaringBitmap<[1, 2, 3]>
Transaction Kind : Pay SUI
Coins:
Object ID : 0xb5eeef175071df6c46efef00b1a56b0fd0d9f8fa
Object ID : 0xdcc4e2369070e03092b9ebb57064c17891a6c84e
Recipient:
0xb92efc6cc3306e52ea8c9e4227fac0470e199adc

----- Transaction Effects ----
Status : Success
Mutated Objects:
  - ID: 0xb5eeef175071df6c46efef00b1a56b0fd0d9f8fa , Owner: Account Address ( 0xb92efc6cc3306e52ea8c9e4227fac0470e199adc )
Deleted Objects:
  - ID: 0xdcc4e2369070e03092b9ebb57064c17891a6c84e

target/debug/sui client pay_sui --recipients 0xb92efc6cc3306e52ea8c9e4227fac0470e199adc --amounts 10000 --input-coins 0x0636d7fa6ddb1900084c0c0a6ea5db868bc9fa06 0x2c2ca2e9cb468e8fff25eea696d669ad22dd4d69 --gas-budget 10000
----- Certificate ----
Transaction Hash: CVZ/bXnne7F6F3cEDUCgK4KYIdwr0Zmlvvm5+GNfqqw=
Transaction Signature: AA==@iEfMV/p70ORu6SACUM3lOaNqKIQdGBLbQQQObHIIbwnmob6rOn2UxmldIXNfYrxZ4T43uSGdcO++ix0UwLS8CA==@xfrRTtxCS07XSV3guhSMkF0a5hw8mIKPPX+LXe7WUZQ=
Signed Authorities Bitmap: RoaringBitmap<[0, 1, 3]>
Transaction Kind : Pay SUI
Coins:
Object ID : 0x0636d7fa6ddb1900084c0c0a6ea5db868bc9fa06
Object ID : 0x2c2ca2e9cb468e8fff25eea696d669ad22dd4d69
Recipients:
0xb92efc6cc3306e52ea8c9e4227fac0470e199adc
Amounts:
10000

----- Transaction Effects ----
Status : Success
Created Objects:
  - ID: 0xcbea8ae4afed2461255745777c1405963a7c7b7d , Owner: Account Address ( 0xb92efc6cc3306e52ea8c9e4227fac0470e199adc )
Mutated Objects:
  - ID: 0x0636d7fa6ddb1900084c0c0a6ea5db868bc9fa06 , Owner: Account Address ( 0xacafa211c1474619c3d2a89a8afcf2612eddf853 )
Deleted Objects:
  - ID: 0x2c2ca2e9cb468e8fff25eea696d669ad22dd4d69

```

### CODE REVIEW

File: crates/sui-json-rpc/src/transaction_builder_api.rs [hunk 1]
```diff
@@ -112,6 +112,35 @@ impl RpcTransactionBuilderServer for FullNodeTransactionBuilderApi {
         Ok(TransactionBytes::from_data(data)?)
     }
 
+    async fn pay_sui(
```
Comment (666lcz) [inline crates/sui-json-rpc/src/transaction_builder_api.rs:115]: Can we add comments about the what these endpoints do as well as all the arguments? Here's a good example:

https://github.com/MystenLabs/sui/blob/b6d56bd22a6076cd4cd1d1cca657da21a6ae2f67/crates/sui-json-rpc/src/api.rs#L208-L223

Comment (gegaowp) [reply crates/sui-json-rpc/src/transaction_builder_api.rs:115]: the corresponding function headers in api.rs have some comments for pay_sui and pay_all_sui, here are implementations of those?

File: crates/sui-json-rpc/src/gateway_api.rs [hunk 2]
```diff
@@ -219,6 +219,35 @@ impl RpcTransactionBuilderServer for TransactionBuilderImpl {
...
+            .pay_sui(signer, input_coins, recipients, amounts, gas_budget)
+            .await?;
+        Ok(TransactionBytes::from_data(data)?)
+    }
+
+    async fn pay_all_sui(
+        &self,
+        signer: SuiAddress,
+        input_coins: Vec<ObjectID>,
+        recipients: SuiAddress,
```
Comment (666lcz) [inline crates/sui-json-rpc/src/gateway_api.rs:241]:
```suggestion
        recipient: SuiAddress,
```

File: crates/sui-json-rpc/src/api.rs [hunk 3]
```diff
@@ -255,6 +255,34 @@ pub trait RpcTransactionBuilder {
         gas_budget: u64,
     ) -> RpcResult<TransactionBytes>;
 
+    #[method(name = "pay_sui")]
```
Comment (666lcz) [inline crates/sui-json-rpc/src/api.rs:258]: Sorry commented in the wrong place before. Can we also add some comments for what this endpoint does on line 241, 258 and 273?

File: crates/sui-json-rpc/src/transaction_builder_api.rs (hunk 1 above)
Comment (666lcz) [reply crates/sui-json-rpc/src/transaction_builder_api.rs:115]: sorry commented in the wrong place. Added a new comment

Comment (gegaowp) [reply crates/sui-json-rpc/src/transaction_builder_api.rs:115]: I am adding some desc for the function itself.

File: crates/sui-json-rpc/src/gateway_api.rs (hunk 2 above)
Comment (gegaowp) [reply crates/sui-json-rpc/src/gateway_api.rs:241]: nice catch thanks!

**Comment by gegaowp** [comment]
updated crates/sui-open-rpc/spec/openrpc.json

File: crates/sui-json-rpc/src/api.rs [hunk 4]
```diff
@@ -238,6 +238,9 @@ pub trait RpcTransactionBuilder {
         amount: Option<u64>,
     ) -> RpcResult<TransactionBytes>;
 
+    /// Send SUI coins to a list of addresses, following a list of amounts,
```
Comment (666lcz) [inline crates/sui-json-rpc/src/api.rs:241]:
```suggestion
    /// Send Coin<T> to a list of addresses, where `T` can be any coin type, following a list of amounts,
```

File: crates/sui-json-rpc/src/api.rs [hunk 5]
```diff
@@ -238,6 +238,9 @@ pub trait RpcTransactionBuilder {
...
+    /// a separate SUI coin object outside of input coins will be used for gas payment,
+    /// if the gas object is not specified, it will be auto selected by the protocol.
```
Comment (666lcz) [inline crates/sui-json-rpc/src/api.rs:243]:
```suggestion
    /// The object specified in the `gas` field will be used to pay the gas fee for the transaction.
    /// The gas object can not appear in `input_coins`. If the gas object is not specified, the RPC server
    /// will auto-select one.
```

**Review by 666lcz — APPROVED** [review]
LGTM! Could you address the comments before merging?

**Comment by gegaowp** [comment]
thanks, resolved!

**Pull request merged**